File: 94_BinaryTreeInorderTraversal/inorderTraversal.py
# 方法一：递归版本

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        f = self.inorderTraversal
        return f(root.left) + [root.val] + f(root.right) if root else []
        
# 方法二：迭代版本

class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        ans = []
        stack = []
        cur = root
        while stack or cur:
            while cur:
                stack.append(cur)
                cur = cur.left
            cur = stack.pop()
            ans.append(cur.val)
            cur = cur.right
        return ans

File: 94_BinaryTreeInorderTraversal/inorderTraversal.go
/**
解法一： 迭代

func inorderTraversal(root *TreeNode) []int {
    stack := []*TreeNode{}
    res := []int{}

    for len(stack) != 0 || root != nil {
        for root != nil {
            stack = append(stack, root)
            root = root.Left
        }

        root = stack[len(stack) - 1]
        // go的语法糖，root.Val 等价 (*root).Val
        // res = append(res, (*root).Val)
        res = append(res, root.Val)
        root = root.Right
        stack = stack[:len(stack) - 1]
    }

    return res
}

 */


/**
解法二：递归
*/
func inorderTraversal(root *TreeNode) []int {
    if root == nil {
        return []int{}
    }
    left := inorderTraversal(root.Left)
    right := inorderTraversal(root.Right)
    return append(append(left, root.Val), right...)
}
